src/line2d.jl

```diff
@@ -1,84 +1,73 @@
-    y0::Number
-    x0::Number
-    y1::Number
-    x1::Number
-end
-
-LineTwoPoints(p1::CartesianIndex{2}, p2::CartesianIndex{2}) = LineTwoPoints(p1[1],p1[2],p2[1],p2[2])
+#Function to return valid intersections of lines with image boundary
 
-immutable LineNormal <: Line
-    ρ::Number
-    θ::Number
+function get_valid_intersections(intersections::Vector{Tuple{Real, Real}}, indsx::Base.OneTo{Int}, indsy::Base.OneTo{Int})
+    valid_intersections = Vector{Tuple{Number,Number}}(0)
+    for intersection in intersections
+        if first(indsx) <= intersection[1] <= last(indsx) && first(indsy) <= intersection[2] <= last(indsy)
+            push!(valid_intersections,intersection)
+        end
+    end
+    valid_intersections
 end
 
-LineNormal(τ::Tuple{Number,Number}) = LineNormal(τ...)
+# LineTwoPoints methods
 
-"""
-```
-img_with_line = line(img, l, color, method)
-img_with_line = line(img, LineTwoPoints(p1, p2), color, method)
-img_with_line = line(img, LineTwoPoints(y0, x0, y1, x1), color, method)
-img_with_line = line(img, LineNormal(ρ,θ), color, method)
-```
+LineTwoPoints(x0::Int, y0::Int, x1::Int, y1::Int) = LineTwoPoints(Point(x0, y0), Point(x1,y1))
+LineTwoPoints(p1::CartesianIndex{2}, p2::CartesianIndex{2}) = LineTwoPoints(Point(p1), Point(p2))
 
-Draws a line on the input image.
-img     =   Image to draw lines on
-l       =   A `Line` type object describing the line to be drawn.
-            A line can be represented in "two-points form" using `LineTwoPoints(p1,p2)`
-            where points p1, p2 are `CartesianIndex{2}`.
-            It can also be represented in "two-points form" using `LineTwoPoints(y0,x0,y1,x1)`
-            A line can also be represented in "normal form" using `LineNormal(ρ,θ)`
-            where ρ = perpendicular distance of line and θ is angle from x-axis in radians.
-color   =   Color of the line to be drawn
-method  =   Lines are drawn using the `bresenham` method by default. If anti-aliasing
-            is required, the `xiaolin_wu` can be used.
+draw!{T<:Colorant}(img::AbstractArray{T,2}, line::LineTwoPoints, method::Function = bresenham) =
+    draw!(img, line, one(T), method)
 
-"""
-
-#methods for LineTwoPoints parametrization
-
-line{T<:Colorant}(img::AbstractArray{T, 2}, args...) = line!(copy(img), args...)
-
-line!{T<:Colorant}(img::AbstractArray{T, 2}, p1::CartesianIndex{2}, p2::CartesianIndex{2}, method::Function = bresenham) =
-    line!(img, LineTwoPoints(p1, p2), one(T), method)
-
-line!{T<:Colorant}(img::AbstractArray{T, 2}, p1::CartesianIndex{2}, p2::CartesianIndex{2}, color::T, method::Function = bresenham) =
-    line!(img, LineTwoPoints(p1, p2), color, method)
+function draw!{T<:Colorant}(img::AbstractArray{T,2}, line::LineTwoPoints, color::T, method::Function = bresenham)
+    indsy, indsx = indices(img)
+    x1 = line.p1.x; y1 = line.p1.y
+    x2 = line.p2.x; y2 = line.p2.y
+    m = (y2 - y1) / (x2 - x1)
+    intersections_x = [(x, y1 + m*(x-x1)) for x in (first(indsx), last(indsx))]
```

Comment:
In case of vertical line `m = Inf`, so `intersections_x = [(first(indsx),Inf), (last(indsx),Inf)]` and `intersections_y = [(x1, first(indsy)),(x1, last(indsy)]`. `valid_intersections = [(x1, first(indsy)),(x1, last(indsy)]` and so the line would be drawn out correctly.
```julia
julia> imshow(draw(zeros(Gray,400,400), LineTwoPoints(200,10,200,20)))
```
![img5](https://cloud.githubusercontent.com/assets/15063205/25047969/81c5cfaa-2157-11e7-92af-6046d26bfb6f.png)